libs/vkd3d/state.c

```diff
@@ -3538,19 +3565,99 @@ static HRESULT vkd3d_bindless_state_add_binding(struct vkd3d_bindless_state *bin
+    }
+
     if ((vr = VK_CALL(vkCreateDescriptorSetLayout(device->vk_device,
             &vk_set_layout_info, NULL, &set_info->vk_host_set_layout))) < 0)
         ERR("Failed to create descriptor set layout, vr %d.\n", vr);
 
     return hresult_from_vk_result(vr);
 }
 
+static bool vkd3d_bindless_supports_mutable_type(struct d3d12_device *device, uint32_t flags)
+{
+    VkDescriptorType descriptor_types[VKD3D_MAX_MUTABLE_DESCRIPTOR_TYPES];
+    const struct vkd3d_vk_device_procs *vk_procs = &device->vk_procs;
+    VkDescriptorSetLayoutBindingFlagsCreateInfoEXT binding_flags;
+    VkMutableDescriptorTypeCreateInfoVALVE mutable_info;
+    VkDescriptorSetLayoutCreateInfo set_layout_info;
+    VkMutableDescriptorTypeListVALVE mutable_list;
+    VkDescriptorSetLayoutSupport supported;
+    VkDescriptorSetLayoutBinding binding;
+
+    VkDescriptorBindingFlagsEXT binding_flag =
+            VK_DESCRIPTOR_BINDING_UPDATE_AFTER_BIND_BIT_EXT |
+            VK_DESCRIPTOR_BINDING_VARIABLE_DESCRIPTOR_COUNT_BIT_EXT |
+            VK_DESCRIPTOR_BINDING_PARTIALLY_BOUND_BIT_EXT |
+            VK_DESCRIPTOR_BINDING_UPDATE_UNUSED_WHILE_PENDING_BIT_EXT;
+
+    if (!device->device_info.mutable_descriptor_features.mutableDescriptorType)
+        return false;
+
+    mutable_info.sType = VK_STRUCTURE_TYPE_MUTABLE_DESCRIPTOR_TYPE_CREATE_INFO_VALVE;
+    mutable_info.pNext = NULL;
+    mutable_info.pMutableDescriptorTypeLists = &mutable_list;
+    mutable_info.mutableDescriptorTypeListCount = 1;
+
+    mutable_list.descriptorTypeCount = vkd3d_bindless_build_mutable_type_list(descriptor_types, flags);
+    mutable_list.pDescriptorTypes = descriptor_types;
+
+    binding.binding = 0;
+    binding.descriptorType = VK_DESCRIPTOR_TYPE_MUTABLE_VALVE;
+    binding.descriptorCount = d3d12_max_descriptor_count_from_heap_type(D3D12_DESCRIPTOR_HEAP_TYPE_CBV_SRV_UAV);
+    binding.pImmutableSamplers = NULL;
+    binding.stageFlags = VK_SHADER_STAGE_ALL;
+
+    binding_flags.sType = VK_STRUCTURE_TYPE_DESCRIPTOR_SET_LAYOUT_BINDING_FLAGS_CREATE_INFO_EXT;
+    binding_flags.pNext = &mutable_info;
+    binding_flags.bindingCount = 1;
+    binding_flags.pBindingFlags = &binding_flag;
+
+    set_layout_info.sType = VK_STRUCTURE_TYPE_DESCRIPTOR_SET_LAYOUT_CREATE_INFO;
+    set_layout_info.pNext = &binding_flags;
+    set_layout_info.flags = VK_DESCRIPTOR_SET_LAYOUT_CREATE_UPDATE_AFTER_BIND_POOL_BIT;
+    set_layout_info.bindingCount = 1;
+    set_layout_info.pBindings = &binding;
+
+    supported.sType = VK_STRUCTURE_TYPE_DESCRIPTOR_SET_LAYOUT_SUPPORT;
+    supported.pNext = NULL;
+    VK_CALL(vkGetDescriptorSetLayoutSupport(device->vk_device, &set_layout_info, &supported));
+    if (!supported.supported)
+        return false;
+
+    set_layout_info.flags = VK_DESCRIPTOR_SET_LAYOUT_CREATE_HOST_ONLY_POOL_BIT_VALVE;
+    binding_flag = VK_DESCRIPTOR_BINDING_VARIABLE_DESCRIPTOR_COUNT_BIT_EXT;
+    VK_CALL(vkGetDescriptorSetLayoutSupport(device->vk_device, &set_layout_info, &supported));
```

Comment:
If it's guaranteed by spec then we should probably revert this since we now have some inconsistent checks (host-only checks for `if (device->device_info.mutable_descriptor_features.mutableDescriptorType)` but the new code suggests that this might in fact not be enough).

Reply:
Changed the checks to use the MUTABLE bindless flag.